### PR TITLE
refactor(api): authorize explicitly

### PR DESCRIPTION
**Pull request by tpluscode**
Related to a discussion with @sandhose, I propose not to secure the entire API by default and instead apply scope-based restrictions on specific handlers/classes

### CODE REVIEW

File: apis/core/hydra/index.ttl [hunk 1]
```diff
@@ -25,6 +31,10 @@ hydra:Resource
     ] .
 
 hydra:Collection
+  auth:required
+    true ;
+  auth:scopes
+    scopes:pipelinesRead ;
```
Comment (tpluscode) [inline apis/core/hydra/index.ttl]: Reading resources and collections would require the `pipelines:read` scope

File: apis/core/hydra/projects.ttl [hunk 2]
```diff
@@ -27,4 +31,6 @@ cc:ProjectsCollection
           code:link
             <file:handlers/cube-projects#post> ;
         ] ;
+      auth:scopes
+        scopes:pipelinesWrite
```
Comment (tpluscode) [inline apis/core/hydra/projects.ttl]: While and request to write resources might require `pipelines:write`

File: apis/core/lib/auth.ts [hunk 3]
```diff
@@ -40,13 +40,13 @@ function devAuthHandler(req: Request, res: Response, next: NextFunction) {
...
-    const permissionHeader = req.headers['x-permission']
-    const permissions = typeof permissionHeader === 'string' ? permissionHeader.split(',').map(s => s.trim()) : permissionHeader || []
+    const permissionHeader = req.headers['x-scope']
+    const scope = typeof permissionHeader === 'string' ? permissionHeader.split(',').map(s => s.trim()) : permissionHeader || []
 
     req.user = {
       id: $rdf.namedNode(sub),
       sub,
-      permissions,
+      scope,
```
Comment (tpluscode) [inline apis/core/lib/auth.ts:49]: `scope` not `permissions` claim used to authorise

File: apis/core/hydra/index.ttl [hunk 4]
```diff
@@ -8,9 +10,12 @@
     <> ;
   hydra:supportedClass
     hydra:Resource,
-    hydra:Collection .
+    hydra:Collection
+.
 
 hydra:Resource
+  auth:scopes
+    scopes:pipelines-read ;
```
Comment (tpluscode) [inline apis/core/hydra/index.ttl:18]: Adding this to a class restricts access to tokens with specific scope(s).

File: apis/core/hydra/projects.ttl [hunk 5]
```diff
@@ -27,4 +31,6 @@ cc:ProjectsCollection
           code:link
             <file:handlers/cube-projects#post> ;
         ] ;
+      auth:scopes
+        scopes:pipelines-write
```
Comment (lucafurrer) [inline apis/core/hydra/projects.ttl:35]: Does the creation of a project require the ability to write to the pipelines or could this be seperated?

Comment (tpluscode) [reply apis/core/hydra/projects.ttl:35]: Certainly. The entire permission is up for discussion

Comment (lucafurrer) [reply apis/core/hydra/projects.ttl:35]: I was not sure, if you had already some discussions.

**Comment by martinmaillard** [comment]
Does that mean anyone could query some endpoints (e.g. api docs), or do you still need a valid token?

Does it make anything more complex?

**Comment by tpluscode** [comment]
Yes, this will open api docs.

Complex? Maybe. The question is how fine-grained we want our access control...

**Comment by martinmaillard** [comment]
That's why I'm asking. We don't need any access control for the MVP. I'm fine with having it already if it doesn't make everything else more complex, as we're probably going to need it later in the project. But if it slows down development significantly, I would drop it for now.

**Comment by tpluscode** [comment]
Fair enough. Closing for now

**Pull request closed without merging**